WIP: Issue 1807 Add Register broker command to Svcat

jberkhahn opened the pull request:
#1807 

This command will allow users to add a broker to service catalog

WIP/ do not merge

luksa commented on cmd/svcat/broker/register_cmd.go:
```diff
@@ -0,0 +1,74 @@
+		Use:   "register NAME --url URL",
+		Short: "Registers a new broker with service catalog",
+		Example: command.NormalizeExamples(`
+		svcat register mysqlbroker --url http://mysqlbroker.com
```
I believe this needs to be `svcat register broker mysqlbroker ...`, so it's aligned with `svcat sync broker mysqlbroker`

jberkhahn replied:
I'm not sure about that. I would thing `register` would be like `provision` and `bind`, neither of which require you to specify the object types they create. @carolynvs thoughts?

carolynvs replied:
I'm with Jonathan on that, for our domain specific verbs that only apply to a single resource type (in this case a broker), I would prefer that we are consistent and not require the redundant noun.

jberkhahn commented:
Ok, this should be ready to be merged now. 

Most of the size is from junk I had to rewrite to turn pkg/svcat/service-catalog into an interface. Everything it does is still the same, it's just specified in an interface so it can be faked out in cmd/svcat for the unit tests.

MHBauer commented on pkg/svcat/service-catalog/sdk.go:
```diff
@@ -17,12 +17,66 @@ limitations under the License.
+	RetrieveSecretByBinding(*apiv1beta1.ServiceBinding) (*apicorev1.Secret, error)
+
+	ServerVersion() (*version.Info, error)
+}
```
Not being sure how this interface is going to be used, I wonder if it best split up into multiple sub interfaces, approximately how it is split by newlines.

jberkhahn replied:
And then what? Have the Svcat app type inline multiple interfaces? Seems like it would be more of a mess to fake, esp. considering you would have to update them individually with Counterfeiter. 

Having one bighuge interface does seem kind of messy, but I don't think splitting it up is actually a better alternative.

MHBauer replied:
Exactly so. If this isn't intended to be used as a nicer version of a client in any way, then we can leave it, otherwise I think it would be nice to have a binder and a provisioner, or a broker client and a plan client, etc.
I'm fine with it as is. Like I said, I haven't thought about how the interface is to be used.

carolynvs replied:
I am fine with a mega interface for now. If there's a need to break it down further, let's tackle that then.

jberkhahn commented on pkg/svcat/svcat.go:
```diff
@@ -24,8 +24,8 @@ import (
-	*servicecatalog.SDK
-
+	//*servicecatalog.SDK
+	SDK servicecatalog.SvcatAPI
```
Other than my dislike of inlining in general? Not really I suppose. I'll revert that bit.

MHBauer commented on pkg/svcat/service-catalog/binding.go:
```diff
@@ -241,16 +241,16 @@ func (sdk *SDK) WaitForBinding(ns, name string, interval time.Duration, timeout
+	return sdk.bindingHasStatus(binding, v1beta1.ServiceBindingConditionFailed)
 }
 
 // BindingHasStatus returns if the instance is in the specified status.
```
`returns true`

jberkhahn replied:
i didn't write that, but thanks

MHBauer commented on pkg/svcat/service-catalog/broker.go:
```diff
@@ -55,6 +55,26 @@ func (sdk *SDK) RetrieveBrokerByClass(class *v1beta1.ClusterServiceClass,
+			},
+		},
+	}
+	result, err := sdk.ServiceCatalog().ClusterServiceBrokers().Create(request)
```
is auth and stuff handled inside of here or is it future work?

jberkhahn commented:
/retest

MHBauer commented on Makefile:
```diff
@@ -216,13 +216,14 @@ verify: .init verify-generated verify-client-gen verify-docs verify-vendor
 	    | grep -v ^pkg/kubernetes/ \
 	    | grep -v generated \
 	    | grep -v ^pkg/client/ \
+	    | grep -v pkg/svcat/service-catalog/fakes \
```
How hard is it to regenerate the fake? This shouldn't be required as of https://github.com/maxbrunsfeld/counterfeiter/pull/66 http://gopkg.in/maxbrunsfeld/counterfeiter.v5

jberkhahn replied:
you just run `counterfeiter . SvcatAPI` in the pkg/svcat/service-catalog directory. That line is needed otherwise the linter complains about the lack of comments in the auto generated code.

MHBauer replied:
I get super different output. Can you try with http://gopkg.in/maxbrunsfeld/counterfeiter.v5 ? It should remove the need to add this extra ignore.

jberkhahn replied:
Yeah I didn't realize my counterfeiter was out of date. Updated it and regenerated the fake.

MHBauer replied:
take the ignore out, and it should work.

MHBauer commented on pkg/svcat/service-catalog/broker_test.go:
```diff
@@ -118,6 +118,41 @@ var _ = Describe("Broker", func() {
+		})
+	})
 	Describe("Sync", func() {
 		It("Useds the generated b1beta1 Retrieve method to get the broker, and then updates it with a new RelistRequests", func() {
```
not your typo `b1beta1`

MHBauer commented on pkg/svcat/service-catalog/service-catalogfakes/fake_svcat_api.go:
```diff
@@ -0,0 +1,2641 @@
+// Code generated by counterfeiter. DO NOT EDIT.
```
perfect.

MHBauer commented on pkg/svcat/svcat.go:
```diff
@@ -24,16 +24,16 @@ import (
 type App struct {
-	*servicecatalog.SDK
-
+	//*servicecatalog.SDK
```
remove comment

MHBauer commented on Makefile:
```diff
@@ -216,13 +216,14 @@ verify: .init verify-generated verify-client-gen verify-docs verify-vendor
 	@#
 	@echo Running golint and go vet:
-	@# Exclude the generated (zz) files for now, as well as defaults.go (it
+	@# Exclude the generated (zz) files in pkg and svcat for now, as well as defaults.go (it
```
also not going to need the comment.

carolynvs commented on pkg/svcat/service-catalog/sdk.go:
```diff
@@ -17,12 +17,68 @@ limitations under the License.
 )
 
+// SvcatAPI is an interface containing the variuos actions in the svcat pkg lib
+// This interface is then faked with Counterfeiter for the cmd/svcat unit tests
```
Please add a blurb to `docs/devguide.md` explaining how to install and use Counterfeiter to perform this task.

carolynvs commented on cmd/svcat/broker/register_cmd_test.go:
```diff
@@ -0,0 +1,95 @@
+			Expect(*cmd).NotTo(BeNil())
+			Expect(cmd.Use).To(Equal("register NAME --url URL"))
+			Expect(cmd.Short).To(Equal("Registers a new broker with service catalog"))
+			Expect(cmd.Example).To(Equal("  svcat register mysqlbroker --url http://mysqlbroker.com"))
```
Can we switch from Equals to Contains so that we aren't so whitespace sensitive?

carolynvs commented on pkg/svcat/service-catalog/binding.go:
```diff
@@ -239,18 +239,18 @@ func (sdk *SDK) WaitForBinding(ns, name string, interval time.Duration, timeout
 
 // BindingHasStatus returns if the instance is in the specified status.
-func (sdk *SDK) BindingHasStatus(binding *v1beta1.ServiceBinding, status v1beta1.ServiceBindingConditionType) bool {
+func (sdk *SDK) bindingHasStatus(binding *v1beta1.ServiceBinding, status v1beta1.ServiceBindingConditionType) bool {
```
👍

carolynvs commented on pkg/svcat/service-catalog/sdk.go:
```diff
@@ -17,12 +17,68 @@ limitations under the License.
 
+// SvcatAPI is an interface containing the variuos actions in the svcat pkg lib
+// This interface is then faked with Counterfeiter for the cmd/svcat unit tests
+type SvcatAPI interface {
```
Can we come up with a different name for this that doesn't include "API"? The reason I went with SDK on the wrapper that I created was that it really wasn't a blind api client like the regenerated go client already is.

jberkhahn commented:
@MHBauer @carolynvs 

Moved some things around a bit. Renamed the pkg/svcat interface to SvcatClient, because it's Svcat's wrapper around the generated client. Is that a better name?

Added a blurb about counterfeiter to docs/devguide.md as well.

MHBauer commented on docs/devguide.md:
```diff
@@ -310,6 +310,13 @@ Keep in mind that golden files help catch errors when the output unexpectedly ch
+Counterfeiter by running `go get http://github.com/maxbrunsfeld/counterfeiter && go install http://github.com/maxbrunsfeld/countefeiter`.
+Then regenerate the fake with `counterfeiter ./path/to/dir/with/interface InterfaceName`. You may have to manually paste the boilerplate
+comment into the fakes.
+
```
is the version of counterfeiter relevant to mention? I know we had some issues to work through.

MHBauer commented on docs/devguide.md:
```diff
@@ -310,6 +310,13 @@ Keep in mind that golden files help catch errors when the output unexpectedly ch
+Certain tests use fakes generated with [Counterfeiter](http://github.com/maxbrunsfeld/counterfeiter). If you add a method
+to an interface (such as SvcatAPI in pkg/svcat/service-catalog) you may need to regenerate the fake. You can install
+Counterfeiter by running `go get http://github.com/maxbrunsfeld/counterfeiter && go install http://github.com/maxbrunsfeld/countefeiter`.
+Then regenerate the fake with `counterfeiter ./path/to/dir/with/interface InterfaceName`. You may have to manually paste the boilerplate
```
good mention of the boilerplate.

carolynvs commented on pkg/svcat/svcat.go:
```diff
@@ -17,25 +17,25 @@ limitations under the License.
 type App struct {
-	*servicecatalog.SDK
-
+	// SvcatAPI is the interface for performing actions against the service-catalog apiserver
```
`SvcatClient`

carolynvs commented on docs/devguide.md:
```diff
@@ -310,6 +310,13 @@ Keep in mind that golden files help catch errors when the output unexpectedly ch
 
+### Counterfeiter
+Certain tests use fakes generated with [Counterfeiter](http://github.com/maxbrunsfeld/counterfeiter). If you add a method
+to an interface (such as SvcatAPI in pkg/svcat/service-catalog) you may need to regenerate the fake. You can install
```
`SvcatClient` instead of `SvcatAPI`

jberkhahn replied:
Fixed.
